delegated_credential/delegated_credential.go

```diff
@@ -0,0 +1,167 @@
+// Copyright 2018 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package delegated_credential
+
+import (
+	"crypto"
+	"crypto/tls"
+	"crypto/tls/ext"
+	"crypto/x509"
+	"time"
+)
+
+func init() {
+	ext.Register(newTLSExtension(ext.DelegatedCredential))
+}
+
+type ProtocolVersion uint16
+
+const (
+	MaxTTLSeconds   = 60 * 60 * 24 * 7 // Seconds
+	MaxTTL          = time.Duration(MaxTTLSeconds) * time.Nanosecond
+	MaxPublicKeyLen = 1 << 16 // Bytes
+)
+
+// Credential stores the public components of the credential.
+//
+// TODO(cjpatton) Rename DelegatedCredentialParams in the draft.
+type Credential struct {
+
+	// Time in nonsecnds for which the credential is valid. When this data
+	// structure is serialized, this value is converted to a uint32 representing
+	// the duration in seconds.
+	ValidTime time.Duration
+
+	// The public key of the credential.
+	PublicKey crypto.PublicKey
+
+	// The signature scheme corresponding to PublicKey.
+	scheme tls.SignatureScheme
+}
+
+// NewECDSACredential generates an ECDSA key pair for the specified group and
+// returns the secret key and a credential with the public key and validity
+// time.
+func NewCredential(
+	scheme tls.SignatureScheme,
+	validTime time.Duration) (crypto.PrivateKey, *Credential, error) {
+	// TODO(cjpatton)
+	return nil, nil, nil
+}
+
+// IsExpired returns true if and only if the credential has not expired.  The
+// end of the validity interval is defined as the deleagtor certificate's
+// notBefore field plus validTime seconds. This function simply checks that the
+// current time is before the end of the valdity interval.
+func (cred *Credential) IsExpired(start, now time.Time) bool {
+	// TODO(cjpatton)
+	return false
+}
+
+func (cred *Credential) HasValidTTL(start, now time.Time) bool {
+	// TODO(cjpatton)
+	return false
+}
+
+// Marshal encodes a credential as per the spec.
+func (cred *Credential) Marshal() ([]byte, error) {
+	// TODO(cjpatton)
+	return nil, nil
+}
+
+// UnmarshalCredential decodes a credential.
+func UnmarshalCredential(serialized []byte) (*Credential, error) {
+	// TODO(cjpatton)
+	return nil, nil
+}
+
+// Delegator stores the secret key of the delegator.
+//
+// This does not implement crypto.Signer, because this interface only works for
+// PKCS1, PSS, and ECDSA signature schemes. We also want to permit use of EdDSA,
```

Comment:
Just a side comment: +1 for EdDSA. Last time I've tried handshake it was 25% faster than ECDSA/P256 (boringssl on arm)